# 2.2 Values and Data Types

A value is a word/number (fundamental thing) that a program processes. These are referably called objects

In [1]:
print(type("Hello, World!"))
print(type(17))
print("Hello, World")

<class 'str'>
<class 'int'>
Hello, World


There are different types of classes/data types, some are integers, strings or float. A way to figure out the class/data type is a function called type

Not suprisingly, sting belongs to the class str and integers belong to the class int

Python has no such restrictions: Inputs and outputs may be of any type.

Another data type is a float, which are numbers with decimal points.

In [2]:
print(type(3.2))

<class 'float'>


Any numbers with quotations are strings. For example:    
"17" & "3.2", they are strings

In [3]:
print(type("17"))
print(type("3.2"))

<class 'str'>
<class 'str'>


See, they're strings.  
Python can accept enclosed quotes that are the same. For example:  
'I like pie', "I like pie", '''I like pie''',

In [4]:
print(type('This is a string.') )
print(type("And so is this.") )
print(type("""and this.""") )
print(type('''and even this...''') )

<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>


Double quotes can contain single quotes, as in "Bob's pie", and single quoted strings can hvae double quotes inside them, as in 'Bob says "I like pie,"', strings that are enclosed with 3 occurences are called triple quoted settings. Those can contain single/double quotes

In [1]:
print('''"Oh no", she exclaimed, "Ben's bike is broken!"''')

"Oh no", she exclaimed, "Ben's bike is broken!"


Triple quoted strings span multiple lines

In [2]:
print("""This message will span
several lines
of the text.""")

This message will span
several lines
of the text.


It doesn't matter if you use single or double quotes or the three-of-a-kind quotes to surround your strings.

In [3]:
print('This is a string.')
print("""And so is this.""")

This is a string.
And so is this.


So the Python language designers usually chose to surround their strings by single quotes. What do you think would happen if the string already contained single quotes?
When you type a large integer, you might be tempted to use commas between groups of three digits, as in 42,000. This is not a legal integer in Python, but it does mean something else, which is legal:

In [4]:
print(42000)
print(42,000)

42000
42 0


Well, that’s not what we expected at all! Because of the comma, Python chose to treat this as a pair of values. In fact, the print function can print any number of values as long as you separate them by commas. Notice that the values are separated by spaces when they are displayed.

In [5]:
print(42, 17, 56, 34, 11, 4.35, 32)
print(3.4, "hello", 45)

42 17 56 34 11 4.35 32
3.4 hello 45


# 2.3 Type of conversion functions

Sometimes it is necessary to convert values from one type to another. Python provides a few simple functions that will allow us to do that. The functions int, float and str will (attempt to) convert their arguments into types int, float and str respectively. We call these type conversion functions.
The int function can take a floating point number or a string, and turn it into an int. For floating point numbers, it discards the decimal portion of the number - a process we call truncation towards zero on the number line. Let us see this in action:

In [6]:
print(3.14, int(3.14))
print(3.9999, int(3.9999))
print(3.0, int(3.0))
print(-3.999, int(-3.999))
print("2345", int("2345"))
print(17, int(17))        
print(int("23bottles"))

3.14 3
3.9999 3
3.0 3
-3.999 -3
2345 2345
17 17


ValueError: invalid literal for int() with base 10: '23bottles'

The last case shows that a string has to be a syntactically legal number, otherwise you’ll get one of those pesky runtime errors. Modify the example by deleting the bottles and rerun the program. You should see the integer 23.
The type converter float can turn an integer, a float, or a syntactically legal string into a float.

In [7]:
print(float("123.45"))
print(type(float("123.45")))

123.45
<class 'float'>


The type converter str turns its argument into a string. Remember that when we print a string, the quotes are removed. However, if we print the type, we can see that it is definitely str.

In [8]:
print(str(17))
print(str(123.45))
print(type(str(123.45)))

17
123.45
<class 'str'>


# 2.4 Variables

The assignment token, =, should not be confused with equality (we will see later that equality uses the == token). The assignment statement links a name, on the left hand side of the operator, with a value, on the right hand side. This is why you will get an error if you enter:

In [9]:
17 = n

SyntaxError: can't assign to literal (<ipython-input-9-beb598da8e38>, line 1)

A common way to represent variables on paper is to write the name with an arrow pointing to the variable’s value. This kind of figure, known as a reference diagram, is often called a state snapshot because it shows what state each of the variables is in at a particular instant in time. 

In [10]:
message = "What's up, Doc?"
n = 17
pi = 3.14159
print(message)
print(n)
print(pi)

What's up, Doc?
17
3.14159


Now, as you step through the statements, you can see the variables and the values they reference as those references are created.
Variables also have types; again, we can ask the interpreter what they are.

In [11]:
message = "What's up, Doc?"
n = 17
pi = 3.14159
print(type(message))
print(type(n))
print(type(pi))

<class 'str'>
<class 'int'>
<class 'float'>


The type of a variable is the type of the object it currently refers to.
We use variables in a program to “remember” things, like the current score at the football game. But variables are variable. This means they can change over time, just like the scoreboard at a football game. You can assign a value to a variable, and later assign a different value to the same variable.A great deal of programming is about having the computer remember things. For example, we might want to keep track of the number of missed calls on your phone. Each time another call is missed, we will arrange to update or change the variable so that it will always reflect the correct value.

# 2.5 Variable Names and Keywords

Variable names can be arbitrarily long. They can contain both letters and digits, but they have to begin with a letter or an underscore. Although it is legal to use uppercase letters, by convention we don’t. If you do, remember that case matters. Bruce and bruce are different variables.

The underscore character ( _) can also appear in a name. It is often used in names with multiple words, such as my_name or price_of_tea_in_china. There are some situations in which names beginning with an underscore have special meaning, so a safe rule for beginners is to start all names with a letter.
If you give a variable an illegal name, you get a syntax error. In the example below, each of the variable names is illegal.

In [12]:
76trombones = "big parade"
more$ = 1000000
class = "Computer Science 101"

SyntaxError: invalid syntax (<ipython-input-12-738a78e2a131>, line 1)

It turns out that class is one of the Python keywords. Keywords define the language’s syntax rules and structure, and they cannot be used as variable names. Python has thirty-something keywords

and	as	assert	break	class	continue
def	del	elif	else	except	exec
finally	for	from	global	if	import
in	is	lambda	nonlocal	not	or
pass	raise	return	try	while	with
yield	True	False	None	

You might want to keep this list handy. If the interpreter complains about one of your variable names and you don’t know why, see if it is on this list.
Programmers generally choose names for their variables that are meaningful to the human readers of the program — they help the programmer document, or remember, what the variable is used for.

# 2.6 Statments and Expressions

A statement is an instruction that the Python interpreter can execute. We have only seen the assignment statement so far. Some other kinds of statements that we’ll see shortly are while statements, for statements, if statements, and import statements. 
An expression is a combination of values, variables, operators, and calls to functions. Expressions need to be evaluated. If you ask Python to print an expression, the interpreter evaluates the expression and displays the result.

In [13]:
print(1 + 1)
print(len("hello"))

2
5


In [14]:
y = 3.14
x = len("hello")
print(x)
print(y)

5
3.14


Note that when we enter the assignment statement, y = 3.14, only the prompt is returned. There is no value. This is due to the fact that statements, such as the assignment statement, do not return a value. They are simply executed.
On the other hand, the result of executing the assignment statement is the creation of a reference from a variable, y, to a value, 3.14. When we execute the print function working on y, we see the value that y is referring to. In fact, evaluating y by itself results in the same response.

# 2.7 Operators and Operands

Operators are special tokens that represent computations like addition, multiplication and division. The values the operator works on are called operands.

In [15]:
20 + 32
hour - 1
hour * 60 + minute
minute / 60
5 ** 2
(5 + 9) * (15 - 7)


NameError: name 'hour' is not defined

The tokens +, -, and *, and the use of parenthesis for grouping, mean in Python what they mean in mathematics. The asterisk (*) is the token for multiplication, and ** is the token for exponentiation. Addition, subtraction, multiplication, and exponentiation all do what you expect.

In [16]:
print(2 + 3)
print(2 - 3)
print(2 * 3)
print(2 ** 3)
print(3 ** 2)

5
-1
6
8
9


When a variable name appears in the place of an operand, it is replaced with the value that it refers to before the operation is performed. For example, what if we wanted to convert 645 minutes into hours. In Python 3, division is denoted by the operator token / which always evaluates to a floating point result.

In [17]:
minutes = 645
hours = minutes / 60
print(hours)

10.75


What if, on the other hand, we had wanted to know how many whole hours there are and how many minutes remain. To help answer this question, Python gives us a second flavor of the division operator. This version, called integer division, uses the token //. It always truncates its result down to the next smallest integer

In [18]:
print(7 / 4)
print(7 // 4)
minutes = 645
hours = minutes // 60
print(hours)


1.75
1
10


Pay particular attention to the first two examples above. Notice that the result of floating point division is 1.75 but the result of the integer division is simply 1. Take care that you choose the correct flavor of the division operator. If you’re working with expressions where you need floating point values, use the division operator /. If you want an integer result, use //.
The modulus operator, sometimes also called the remainder operator or integer remainder operator works on integers (and integer expressions) and yields the remainder when the first operand is divided by the second. In Python, the modulus operator is a percent sign (%). The syntax is the same as for other operators.

In [19]:
quotient = 7 // 3    
print(quotient)
remainder = 7 % 3
print(remainder)


2
1


The modulus operator turns out to be surprisingly useful. For example, you can check whether one number is divisible by another—if x % y is zero, then x is divisible by y. Also, you can extract the right-most digit or digits from a number. For example, x % 10 yields the right-most digit of x (in base 10). Similarly x % 100 yields the last two digits.
Finally, returning to our time example, the remainder operator is extremely useful for doing conversions, say from seconds, to hours, minutes and seconds. If we start with a number of seconds, say 7684, the following program uses integer division and remainder to convert to an easier form. Step through it to be sure you understand how the division and remainder operators are being used to compute the correct values.

# 2.8 Input

The input function allows the user to provide a prompt string. When the function is evaluated, the prompt is shown. The user of the program can enter the name and press return. When this happens the text that has been entered is returned from the input function, and in this case assigned to the variable n. Make sure you run this example a number of times and try some different names in the input box that appears.

In [20]:
n = input("Please enter your name: ")
print("Hello", n)

Please enter your name: Luke
Hello Luke


It is very important to note that the input function returns a string value. Even if you asked the user to enter their age, you would get back a string like "17". It would be your job, as the programmer, to convert that string into an int or a float, using the int or float converter functions we saw earlier.
To modify our previous program, we will add an input statement to allow the user to enter the number of seconds. Then we will convert that string to an integer. From there the process is the same as before. To complete the example, we will print some appropriate output.

In [21]:
str_seconds = input("Please enter the number of seconds you wish to convert")
total_secs = int(str_seconds)

hours = total_secs // 3600
secs_still_remaining = total_secs % 3600
minutes =  secs_still_remaining // 60
secs_finally_remaining = secs_still_remaining  % 60

print("Hrs=", hours, "mins=", minutes, "secs=", secs_finally_remaining)

Please enter the number of seconds you wish to convert3600
Hrs= 1 mins= 0 secs= 0


The variable str_seconds will refer to the string that is entered by the user. As we said above, even though this string may be 7684, it is still a string and not a number. To convert it to an integer, we use the int function. The result is referred to by total_secs. Now, each time you run the program, you can enter a new value for the number of seconds to be converted.